compiler/rustc_mir_transform/src/dataflow_const_prop.rs

```diff
@@ -0,0 +1,490 @@
+                        FlatSet::Top => FlatSet::Top,
+                        FlatSet::Elem(Const::Scalar(scalar, _)) => {
+                            FlatSet::Elem(scalar.assert_bits(scalar.size()))
+                        }
+                        FlatSet::Bottom => FlatSet::Bottom,
+                    };
+                    discr_value = Some(result);
+                    result
+                }
+            };
+
+            let FlatSet::Elem(choice) = discr_value else {
+                // Do nothing if we don't know which branch will be taken.
+                return
+            };
+
+            if target.value.map(|n| n == choice).unwrap_or(!handled) {
+                // Branch is taken. Has no effect on state.
+                handled = true;
+            } else {
+                // Branch is not taken, we can flood everything.
+                state.flood_all(FlatSet::Bottom);
+            }
+        })
+    }
+}
+
+#[derive(Clone, PartialEq, Eq)]
+enum Const<'tcx> {
+    // FIXME: If there won't be any other cases, make it a struct.
+    Scalar(ScalarInt, Ty<'tcx>),
+}
+
+impl<'tcx> std::fmt::Debug for Const<'tcx> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            Self::Scalar(scalar, ty) => {
+                std::fmt::Display::fmt(&ConstantKind::Val(ConstValue::Scalar(scalar.into()), ty), f)
+            }
+        }
+    }
+}
+
+impl<'tcx> ConstAnalysis<'tcx> {
+    pub fn new(tcx: TyCtxt<'tcx>, body: &Body<'tcx>, map: Map) -> Self {
+        Self {
+            map,
+            tcx,
+            ecx: InterpCx::new(tcx, DUMMY_SP, ty::ParamEnv::empty(), DummyMachine),
+            param_env: tcx.param_env(body.source.def_id()),
+        }
+    }
+
+    fn binary_op(
+        &self,
+        state: &mut State<FlatSet<Const<'tcx>>>,
+        op: BinOp,
+        left: &Operand<'tcx>,
+        right: &Operand<'tcx>,
+    ) -> (FlatSet<Const<'tcx>>, FlatSet<Const<'tcx>>) {
+        let left = self.eval_operand(left, state);
+        let right = self.eval_operand(right, state);
+        match (left, right) {
+            (FlatSet::Elem(left), FlatSet::Elem(right)) => {
+                match self.ecx.overflowing_binary_op(op, &left, &right) {
+                    Ok((val, overflow, ty)) => {
+                        let val = val
+                            .try_to_int()
+                            .ok()
+                            .map(|val| self.wrap_scalar(val, ty))
+                            .unwrap_or(FlatSet::Top);
+                        let overflow = self.wrap_scalar(overflow.into(), self.tcx.types.bool);
+                        (val, overflow)
+                    }
+                    _ => (FlatSet::Top, FlatSet::Top),
+                }
+            }
+            (FlatSet::Bottom, _) | (_, FlatSet::Bottom) => (FlatSet::Bottom, FlatSet::Bottom),
+            (_, _) => {
+                // Could attempt some algebraic simplifcations here.
```

Comment:
The issue is that the simplifications could enable further constant propagation opportunities. We would have to run constant propagation after instruction simplification again, potentially alternating until a fixpoint is reached. So either we handle simplifications here, or we just ignore these optimization opportunities (which would probably also be fine, as this should be quite rare).

Reply:
Yes, but this is a much more general problem and applies to other combinations of optimizations too. I'd want to see some evidence that there is an outsized need here before introducing this.

Reply:
Yes, makes sense. Although I just noticed that the old/current constant propagation also does algebraic simplifications.

Reply:
Yes, it also has the same bug wrt writing to "inactive" enum variants. To be clear, I'm not fundamentally opposed to this. I could even imagine `InstSimplify` exposing an API like `simplify_instruction(&mut Statement)` that you can call on individual instructions. That might make it easier and less dangerous to add this logic here.

Reply:
A `&mut Statement` API would not be sufficient, because the analysis keeps a `&Body`. But yes, there might be an opportunity here. I added a link to this subthread in the initial post.